### PR TITLE
Add "Hot Close" Automatic Persistence of Unsaved Changes Via WebSQL - #10222 -

### PR DESCRIPTION
**Pull request by nyteksf**
_**Topic: Issue #10222 -** "Easy to hit Reload by accident, losing Undo history for files that don't have unsaved changes"_

_Excerpt:_
> If you do reload Brackets, or quit & relaunch, etc., it would be cool if the Undo history were preserved -- or a more coarse-grained 'local history' at least.
This PR has evolved over time. It began as [PR #14413](https://github.com/adobe/brackets/pull/14413); adding general persistence to opened files using Chromium's LocalStorage. This feature was limited to providing only 5MB of storage at the maximum, and also applied to saved files. Then, @nethip suggested limiting the [scope of that new PR](url) to only persisting unsaved changes, so I did this (also accomplished using LocalStorage).

Moving past those steps to get to where I am now, I've now decided to switch to using WebSQL for persisting unsaved changes. I've set it to 500MB capacity, and with automatic overflow protection, although the amount of reserved space is flexible. The code is written to use straight forward SQL queries to persist user changes in the document text, cursor and scroll positions, as well as undo/redo history on keyup events primarily. This data is encoded and compressed before DB storage. When the 'hotClose' preference has been set to 'true', on the event of a single file being closed or when refreshing Brackets via F5, all save file dialogs are skipped; allowing the user to persist from exactly where they'd left off without skipping a beat.

### CODE REVIEW

**Comment by nethip** [comment]
@nyteksf Kudos on this PR! Great job!

Apologies from my side in reviewing this PR. Got caught with other work! Please bear with me on this. I will review this code as soon as I get a chance. Thanks!

**Comment by nyteksf** [comment]
@nethip No worries. :) And thanks, by the way!
I'm looking forward to whatever comments or suggestions you've got.

File: .gitignore
Comment (saulotoledo) [inline .gitignore]: Is this file supposed to be removed?

File: .brackets.json
Comment (saulotoledo) [inline .brackets.json]: Keep the newline in the end of the file.

File: package.json
Comment (saulotoledo) [inline package.json]: Why remove eslint here?

File: package.json
Comment (saulotoledo) [inline package.json]: Keep the line in the end of the file.

File: src/config.json
Comment (saulotoledo) [inline src/config.json]: Why remove eslint here?

File: src/document/DocumentCommandHandlers.js
Comment (saulotoledo) [inline src/document/DocumentCommandHandlers.js]: Trailling whitespace to be removed.

File: src/document/DocumentCommandHandlers.js
Comment (saulotoledo) [inline src/document/DocumentCommandHandlers.js]: Trailing whitespace creating unnecessary diff. There are other cases like this in your commit. I suggest to review them for your next submission.

File: src/document/DocumentCommandHandlers.js
Comment (saulotoledo) [inline src/document/DocumentCommandHandlers.js]: Fix params alignment for the last 2 lines.

File: src/document/DocumentCommandHandlers.js
Comment (saulotoledo) [inline src/document/DocumentCommandHandlers.js]: Why remove the error param here?

File: src/document/DocumentCommandHandlers.js
Comment (saulotoledo) [inline src/document/DocumentCommandHandlers.js]: Why remove the error param here?

File: src/document/DocumentCommandHandlers.js
Comment (saulotoledo) [inline src/document/DocumentCommandHandlers.js]: Fix indent for this code. You are mixing tabs and spaces, use only spaces here. There are other cases like this in your commit that should be fixed.

File: src/editor/Db.js
Comment (saulotoledo) [inline src/editor/Db.js]: I do not believe that it is a good idea just to send the error to the console. If everyone does that, the console will be a bit messy. These errors should be properly threated in the application. As I do not have experience with the platform, I will keep the point to @nethip, but this is just a comment that can help you to think while waiting for a proper review. Note that the same happens in some other areas of your contribution.

File: src/editor/Editor.js
Comment (saulotoledo) [inline src/editor/Editor.js]: Most of these changes are not necessary. Just keep the values aligned as before.

File: src/editor/Editor.js
Comment (saulotoledo) [inline src/editor/Editor.js]: Same comment as for line 117.

File: src/editor/Editor.js
Comment (saulotoledo) [inline src/editor/Editor.js]: Same comment as for line 117. Do the same for the rest of the file.

File: src/editor/Editor.js
Comment (saulotoledo) [inline src/editor/Editor.js]: Most of the changes here are formatting the file, but they are not necessary and will cause trouble for other developers while rebasing their repositories, for example. Just keep it simple and focus in your changes.

**Review by saulotoledo — CHANGES_REQUESTED** [review]
Hi @nyteksf! While you're waiting for @nethip's review, I've added some comments in your code to help in your contribution. I do not have experience with the platform, so the important things will come on @nethip's comments, but I am sure that the changes I suggested will help him in his review. Additionaly, I belieeyou have too many commits. They are ok for you to organize how do you send code to your repository, but once they are merged it is hard to track the commits that are related to a fix or task. I suggest to squash them in a single commit and add a good description (note that you will loose some green squares in your GitHub account if you do that). Another good suggestion is to configure your IDE to proper handle tabs and spaces for indentation. You will notice that mixing them keeps the diffs hard to read. Let me know if you have questions about. Looking forward to see your updated changes! :smile:

**Comment by navch** [comment]
Great stuff @nyteksf 👍 Few comments from my side. Good review from @saulotoledo 

Please revert files which don't have your changes. I am not really sure why you made changes in config.json and package.json I am not able to find where are you using Bluebird. so Please revert unwanted changes. So That will be very easy to review. If you want I can comment what all files/changes that you need to revert from this PR.

File: .gitignore
Comment (nyteksf) [reply .gitignore]: I am admittedly still somewhat new here. Could I ask you if ignoring a file is the same as removing one from a commit? I at least really think I remember someone telling me in my first few commits a while ago to ignore .gitignore, so I had tried using 'git remove <file>' to exclude it from the commit. I'd not been able to get confirmation that it was the right approach, and am still uncertain of what to do only if you have a moment to help me clarify, please.

File: src/config.json
Comment (nyteksf) [reply src/config.json]: I just fixed this now. I was having some trouble running the script a while ago. But I've since gotten everything working, and forgot to enable it again.

File: src/document/DocumentCommandHandlers.js
Comment (nyteksf) [reply src/document/DocumentCommandHandlers.js]: Removed.

**Review by nyteksf — COMMENTED** [review]
@saulotoledo Real quick before beginning, thanks again for taking the time to help me out here. That said, I plan to heed your advice by squashing all of my local commits before I push in each future PR, but meanwhile, would you happen to have any suggestions you could offer on squashing commits that have already been pushed and are in a PR? I've recently experimented on an old branch using [this](https://gist.github.com/longtimeago/f7055aa4c3bba8a62197) set of commands, and also by rebasing locally before pushing, but neither have seemed to do anything to the remote branch or the associated PR, unfortunately. But I definitely see your point about the clutter occurring otherwise.

Otherwise, I wanted to say I've since implemented the changes you've suggested. And I was just wondering, but did you have something else in mind for the method of error handling in Db.js, if I could ask? At present, I went about it through adding a dynamic modal which fires and funnels any arising errors through to the user as they occur. But I'm definitely open to any suggestions.

**Comment by nyteksf** [comment]
@navch Thanks! I look forward to many more PRs here with your help and that coming from everyone else like @saulotoledo. I'm definitely going to attempt once again to fix my master branch I'd forked. It seems I can't delete it without deleting all of my open PRs as well. I mean, Git can sometimes be so confusing for me even still, along with what exactly Brackets needs included or excluded. So I'd managed to mangle the master branch by accident long ago when I altered it and then pushed the first time, and have been reusing an old branch that actually works, and which I proceed to modify anew for each project. 

To explain Bluebird and the other additions, I was experimenting with other database options before settling on WebSQL/SQLite3, and forgot to remove what you've seen. My bad. I'm going to remove them now, but feel free to please let me know if I've missed anything on review. Thank you again in any case, and have a good one.

**Comment by nyteksf** [comment]
@saulotoledo Okay, thank you again for the help. I've made progress, although it wasn't without a bit of stumbling. 

That said, regarding my master branch: I'd followed the instructions for rebasing, and have added back .gitignore as well, but when cloning said repo to my computer and executing brackets, it still hangs almost right after opening. It gets to the title bar saying 'index.html' and then just dies with the main screen remaining white where the editor would appear, basically. Does this mean something else is missing from that branch which is needed to run, maybe?

Regarding the Hot Close commit, next: I cloned down another copy locally for this task. I did rebase that local copy more or less successfully in that I squashed many, many, many commits down into a single one. Although I should add that after two days and a number of hours of merging the same documents over and over again very carefully, I began to burn out and started just merging them with the intent of fixing the formatting before the end. However, I was surprised to discover that the final squashed commit I had pushed back up didn't match my prior pushed changes, and was actually several versions off from being the most recent. I didn't want to play around with breaking something and possibly doing everything over again, so just committed repairs directly to the Hot Close branch through Github. Now that I have some experience, I do feel confident that next time I can prevent the mess from happening a lot easier than I could clean it up. And that's important! So thank you again. What were you saying about the error handling though, if I may, please?

**Comment by nyteksf** [comment]
@saulotoledo Hey again. If you have a moment, please, I'm a bit confused. I re-ran the above commands you'd given once again, and so was able to squash several commits down to one, and then had to add everything remaining to be tracked in order to rebase. That's now done. But apparently the branch has conflicts, and pulling down from upstream didn't give me the ability to rectify them. If I may, how do I solve the above conflicts, please? It doesn't at least seem I have access to merge them locally like with before, so for now I'm kind of stumped.

**Pull request closed without merging**
